modules/calc/src/main/java/com/opengamma/strata/calc/CalculationRunner.java

```diff
@@ -0,0 +1,174 @@
+      List<Column> columns,
+      CalculationRules calculationRules,
+      CalculationEnvironment marketData);
+
+  //-------------------------------------------------------------------------
+  /**
+   * Performs calculations asynchronously for a single scenario,
+   * invoking a listener as each calculation completes.
+   * <p>
+   * This method requires the listener to assemble the results, but it can be much more memory efficient when
+   * calculating aggregate results. If the individual results are discarded after they are incorporated into
+   * the aggregate they can be garbage collected.
+   * 
+   * @param targets  the targets for which values of the measures will be calculated
+   * @param columns  the configuration for the columns that will be calculated, including the measure and
+   *   any column-specific overrides
+   * @param calculationRules  the rules defining how the calculation is performed
+   * @param marketData  market data to be used in the calculations
+   * @param listener  listener that is invoked when individual results are calculated
+   */
+  public abstract void calculateSingleScenarioAsync(
+      List<? extends CalculationTarget> targets,
+      List<Column> columns,
+      CalculationRules calculationRules,
+      CalculationEnvironment marketData,
+      CalculationListener listener);
+
+  /**
+   * Performs calculations asynchronously for a multiple scenarios, each with a different set of market data,
+   * invoking a listener as each calculation completes.
+   * <p>
+   * This method requires the listener to assemble the results, but it can be much more memory efficient when
+   * calculating aggregate results. If the individual results are discarded after they are incorporated into
+   * the aggregate they can be garbage collected.
+   * 
+   * @param targets  the targets for which values of the measures will be calculated
+   * @param columns  the configuration for the columns that will be calculated, including the measure and
+   *   any column-specific overrides
+   * @param calculationRules  the rules defining how the calculation is performed
+   * @param marketData  the market data used in the calculations
+   * @param listener  listener that is invoked when individual results are calculated
+   */
+  public abstract void calculateMultipleScenariosAsync(
+      List<? extends CalculationTarget> targets,
+      List<Column> columns,
+      CalculationRules calculationRules,
+      CalculationEnvironment marketData,
+      CalculationListener listener);
+
+  //-------------------------------------------------------------------------
+  /**
+   * Gets the underlying task runner.
+   * <p>
+   * In most cases, this runner will be implemented using an instance of {@link CalculationTaskRunner}.
+   * That interface provides a lower-level API, with the ability optimize if similar calculations
+   * are being made repeatedly. In addition, the lower level API provides the ability to determine
```

Comment:
The lower level API isn't actually necessary to determine the requirements, you only need to create an instance of `CalculationTasks` and call `getRequirements`.